src/segment-loader.js

```diff
@@ -309,6 +309,97 @@ export const shouldWaitForTimelineChange = ({
   return false;
 };
 
+export const mediaDuration = ({ audioTimingInfo, videoTimingInfo }) => {
+  const audioDuration =
+    audioTimingInfo &&
+    typeof audioTimingInfo.start === 'number' &&
+    typeof audioTimingInfo.end === 'number' ?
+      audioTimingInfo.end - audioTimingInfo.start : 0;
+  const videoDuration =
+    videoTimingInfo &&
+    typeof videoTimingInfo.start === 'number' &&
+    typeof videoTimingInfo.end === 'number' ?
+      videoTimingInfo.end - videoTimingInfo.start : 0;
+
+  return Math.max(audioDuration, videoDuration);
+};
+
+export const segmentTooLong = ({ segmentDuration, maxDuration }) => {
+  // 0 duration segments are most likely due to metadata only segments or a lack of
+  // information.
+  if (!segmentDuration) {
+    return false;
+  }
+
+  // For HLS:
+  //
+  // https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#section-4.3.3.1
+  // The EXTINF duration of each Media Segment in the Playlist
+  // file, when rounded to the nearest integer, MUST be less than or equal
+  // to the target duration; longer segments can trigger playback stalls
+  // or other errors.
+  //
+  // For DASH, the mpd-parser uses the largest reported segment duration as the target
+  // duration. Although that reported duration is occasionally approximate (i.e., not
+  // exact), a strict check may report that a segment is too long more often in DASH.
+  return Math.round(segmentDuration) > maxDuration + TIME_FUDGE_FACTOR;
+};
+
+export const getTroublesomeSegmentDurationMessage = ({
+  audioTimingInfo,
+  videoTimingInfo,
+  targetDuration,
+  sourceType,
+  segmentInfo
+}) => {
```

Comment:
why not pass in segmentInfo and sourceType? Since the audio/video timing info and targetDuration are both part of that?
```
({
  audioTimingInfo,
  videoTimingInfo,
  mediaIndex,
  duration,
  playlist: { id, targetDuration }
}, sourecType) => {}
```
with the following below
```
getTroublesomeSegmentDurationMessage(segmentInfo, this.sourceType_);
```